File: leetcode/src/00042_trapping_rain_water/solution.go
package leetcode

func trap(height []int) int {
	dp1 := make([]int, len(height))
	dp1[0] = height[0]
	for i := 1; i < len(height); i++ {
		dp1[i] = max(dp1[i-1], height[i])
	}
	dp2 := make([]int, len(height))
	dp2[len(height)-1] = height[len(height)-1]
	for i := len(height) - 2; i >= 0; i-- {
		dp2[i] = max(dp2[i+1], height[i])
	}
	res := 0
	for i := 0; i < len(height); i++ {
		res += min(dp1[i], dp2[i]) - height[i]
	}
	return res
}

func max(x int, y int) int {
	if x < y {
		return y
	}
	return x
}

func min(x int, y int) int {
	if x < y {
		return x
	}
	return y
}

File: leetcode/src/00042_trapping_rain_water/solution.rs
struct Solution;

impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        let mut res = 0;
        let mut st = Vec::new();
        for (right, &current_h) in height.iter().enumerate() {
            while !st.is_empty() && current_h > height[*st.last().unwrap()] {
                let lower_bound_h = height[st.pop().unwrap()];
                if st.is_empty() {
                    break;
                }
                let left = *st.last().unwrap();
                let width = (right - left - 1) as i32;
                let upper_bound_h = current_h.min(height[left]);
                res += width * (upper_bound_h - lower_bound_h);
            }
            st.push(right);
        }
        res
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        let height = vec![0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1];
        assert_eq!(Solution::trap(height), 6);

        let height = vec![4, 2, 0, 3, 2, 5];
        assert_eq!(Solution::trap(height), 9);

        let height = vec![2, 1, 0, 0, 0, 0, 0, 1, 2];
        assert_eq!(Solution::trap(height), 12);
    }
}
